hotshot-state-prover/src/service.rs

```diff
@@ -58,40 +62,83 @@ pub struct StateProverConfig {
-    (0..num_node).for_each(|id| {
-        let bls_key = BLSPubKey::generated_from_seed_indexed(seed, id as u64).0;
-        let state_ver_key = StateKeyPair::generate_from_seed_indexed(seed, id as u64).ver_key();
-        st.register(bls_key, U256::from(1u64), state_ver_key)
-            .expect("Key registration shouldn't fail.");
-    });
+    st.batch_register(
+        bls_keys.iter().cloned(),
+        iter::repeat(U256::one()).take(bls_keys.len()),
+        state_keys.iter().cloned(),
+    )?;
     st.advance();
     st.advance();
-    st
+    Ok(st)
 }
 
-pub async fn init_stake_table_from_config(
-    config: &StateProverConfig,
+pub async fn init_stake_table_from_orchestrator(
+    orchestrator_url: &Url,
 ) -> StakeTable<BLSPubKey, StateVerKey, CircuitField> {
-    let st = init_stake_table(config.num_nodes, config.seed);
-    std::println!("Stake table initialized.");
-    st
+    tracing::info!("Initializing stake table from HotShot orchestrator.");
+    let client = Client::<ServerError>::new(orchestrator_url.clone());
+    let mut num_retries = 10;
+    while num_retries > 0 {
+        match client.get::<bool>("api/peer_pub_ready").send().await {
+            Ok(true) => {
+                match client
+                    .get::<NetworkConfig>("api/config_after_peer_collected")
+                    .send()
+                    .await
+                {
+                    Ok(config) => {
+                        let mut st = StakeTable::<BLSPubKey, StateVerKey, CircuitField>::new(
+                            STAKE_TABLE_CAPACITY,
+                        );
+                        config
+                            .config
+                            .known_nodes_with_stake
+                            .into_iter()
+                            .for_each(|config| {
+                                st.register(
+                                    *config.stake_table_entry.get_key(),
+                                    config.stake_table_entry.get_stake(),
+                                    config.state_ver_key,
+                                )
+                                .expect("Key registration shouldn't fail.");
+                            });
+                        st.advance();
+                        st.advance();
+                        return st;
+                    }
+                    Err(e) => {
+                        num_retries -= 1;
+                        tracing::warn!("Orchestrator error: {e}, retrying.");
+                    }
+                }
+            }
+            Ok(false) => {
+                num_retries -= 1;
+                tracing::info!("Peers' keys are not ready.");
+            }
+            Err(e) => {
+                num_retries -= 1;
+                tracing::warn!("Orchestrator error {e}, retrying.");
+            }
+        }
+        sleep(Duration::from_secs(2)).await;
+    }
+    panic!("Failed initializing stake table.")
```

Comment:
I think that this can retry indefinitely, don't see why we need a retry limit here

Reply:
changed in [43e59c7](https://github.com/EspressoSystems/espresso-sequencer/pull/1183/commits/43e59c71050e5be3330952d4b273289699775292)